### ISSUE TITLE
glmReserve

### ISSUE THREAD

**Issue by grishaloshadinov** [mages/ChainLadder#2]
when I run "glmReserve" in RExcel with "var.power=1" and "cum=FALSE" and "mse.method = bootstarp" and "nsim =1000" I get error message "Microsoft Excel is waiting for another application to complete an OLE action"
When I run it in R it takes absolutely ages and eventually I have to kill it with getting no results

**Comment by trinostics** [comment]
It would appear that your RExcel issue and your "R it takes absolutely ages" issue are related. Please provide a minimal, reproducible example **without** RExcel. If you cannot reproduce the issue **without** RExcel, please resolve your issue with the RExcel people.

**Comment by grishaloshadinov** [comment]
I can't reproduce it in R either. How can I attach excel file with raw triangle that I am using please?

**Comment by Rick132** [comment]
Hello!
I have the same problem. I am working in R. When I run it in R it takes absolutely ages and eventually I have to kill it with getting no results. I hope someone can help

**Comment by actuaryzhang** [comment]
Is there an example you can provide to reproduce the issue? Is this issue with glmReserve or RExcel?

**Comment by grishaloshadinov** [comment]
that's what I mean, they didn't take me seriously, more flaws will follow 

    On Monday, 25 January 2016, 17:12, actuaryzhang <notifications@github.com> wrote:
 

 Is there an example you can provide to reproduce the issue? Is this issue with glmReserve or RExcel? —

**Comment by Rick132** [comment]
Hello there!
Thanks for replying. I am using glmReserve in R. Find attached the file. I am not sure if it because the incremental payment contains some zeros. Also, is there a way I can extract the process and parameter risk from glmReserve either with analytical or bootstrapping?  Do the package provide RMSEP of Munich Chain Ladder? Hoping to get a response soon.
[glmreserve1.pdf](https://github.com/mages/ChainLadder/files/105915/glmreserve1.pdf)

**Comment by actuaryzhang** [comment]
Thank you very much for providing the example. It seems that the resampling generates lots of negative values of the incrementals, which are not allowed in the current implementation. The reason that it takes forever is because it keeps resampling until all values are nonnegative. This criterion is an intuitive but perhaps not the best way to handle negative payments. I will investigate this problem more and try to use a better approach. Hopefully I can solve this soon but no promise given my current schedule. I will keep you posted. 

Right now, the glmReserve function does not report separately the process and estimation errors. It is available, however, in the code. I can make the function report these details in the next release, but I seldom find such a breakdown meaningful.

**Comment by Rick132** [comment]
Thanks for the reply.

**Comment by kennedymwavu** [comment]
Has anyone found workarounds for this issue? I'm facing the same problem in R.

**Comment by actuaryzhang** [comment]
Hi Dan, good to hear from you. It's been a long time since the last time I
worked on this. IIRC, the issue is fitting the GLM on data with negative
values generated from the simulation.
One workaround is setting negative values to zero (and redistributing the
value to other cells).  Feel free to make this change.

Wayne

**Comment by trinostics** [comment]
@actuaryzhang In another context I have chosen to "wipe out" negative simulated values when a subsequent dependent calculation / function requires positive support (like 'log'). Is this a similar situation? What about replacing the negative values with their absolute values, rather than with zero? If replace with zero, what do you mean by "redistributing  the value to other cells"? 

Thus uploaded R script (renamed 'txt' for uploading) contains a few small functions that came in handy with that previous context, perhaps useful here. 
[truncated_normal.txt](https://github.com/mages/ChainLadder/files/8326795/truncated_normal.txt)

@grishaloshadinov @Rick132 @kennedymwavu Do you have any interest in working out a solution to this issue? I am not a user of glmReserve so "I have no horse in this race." If you do not pick up this thread, it will probably remain dormant. But if you do, I will gladly watch and comment.

Dan

**Comment by kennedymwavu** [comment]
@trinostics Yes, I'd be glad. But I have no knowledge of claims reserving so I'm not sure if I can be of help.

**Comment by trinostics** [comment]
@kennedymwavu I targeted you in my reply above because, when I saw your comment [Has anyone found workarounds for this issue? I'm facing the same problem in R.], I assumed you had a reserving-related reason to ask. IMO, the two of us without a horse in the race would not likely spearhead a successful solution. Where does your interest originate? More in the glm/bootstrap theory? @grishaloshadinov , as the original poster (OP), are you interested in helping see this through? How about you, @Rick132 [Hello! I have the same problem.] ? If no one is interested in this issue anymore, then it should be closed.

**Comment by udaykumarravada** [comment]
Sorry for if I miss any information. How can we apply a GLM to negative incremental values i tried for small triangle of 10 X 10 with few negative values I have translated them made the negative values to zero. 
But now the triangle is very large which 45 X 45 and has a lot of negative values and what should I do?
Any help would be very appreciated?

**Comment by udaykumarravada** [comment]
> I think you're trying to force a square peg in a round hole. Consider a different model that is more applicable to your data.

Thanks for the reply.
Can you please suggest some other  models ?

**Comment by jdonland** [comment]
It might be helpful to mention in the package documentation which modelling procedures tolerate negative loss increments and which don't.

A cursory look suggests that Mack belongs to the former category while Clark and GLM belong to the latter. 

Negative loss increments can be genuinely useful in practice (e.g. when modelling the release of allocated reserves) so it's not too surprising that users fall into this trap.
